PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc

```diff
@@ -30,59 +30,61 @@
-                return - dx * std::sin(phi) + dy * std::cos(phi);
-            }
-            float calcDz(reco::Candidate::Point p, reco::Candidate::Point v, const reco::Candidate &c) {
-                return p.Z()-v.Z() - ((p.X()-v.X()) * c.px() + (p.Y()-v.Y())*c.py()) * c.pz()/(c.pt()*c.pt());
-            }
-    };
+  ///conversion map from quality flags used in PV association and miniAOD one
+  namespace {
+    const static int qualityMap[8]  = {1,0,1,1,4,4,5,6};
+  }
+  
+  class PATPackedCandidateProducer : public edm::global::EDProducer<> {
+  public:
+    explicit PATPackedCandidateProducer(const edm::ParameterSet&);
+    ~PATPackedCandidateProducer();
+    
+    virtual void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const;
+    
+    //sorting of cands to maximize the zlib compression
+    bool candsOrdering(pat::PackedCandidate i,pat::PackedCandidate j) const {
+      if (std::abs(i.charge()) == std::abs(j.charge())) {
+        if(i.charge()!=0){
+          if(i.pt() > minPtForTrackProperties_ and j.pt() <= minPtForTrackProperties_ ) return true;
+          if(i.pt() <= minPtForTrackProperties_ and j.pt() > minPtForTrackProperties_ ) return false;
+        }
+        if(i.vertexRef() == j.vertexRef()) 
+          return i.eta() > j.eta();
+        else 
+          return i.vertexRef().key() < j.vertexRef().key();
+      }
+      return std::abs(i.charge()) > std::abs(j.charge());
+    }
+    template <typename T>
+    std::vector<size_t> sort_indexes(const std::vector<T> &v ) const {
+      std::vector<size_t> idx(v.size());
+      for (size_t i = 0; i != idx.size(); ++i) idx[i] = i;
+      std::sort(idx.begin(), idx.end(),[&v,this](size_t i1, size_t i2) { return candsOrdering(v[i1],v[i2]);});
+      return idx;
+    }
+    
+  private:
+    const edm::EDGetTokenT<reco::PFCandidateCollection>    Cands_;
+    const edm::EDGetTokenT<reco::VertexCollection>         PVs_;
+    const edm::EDGetTokenT<edm::Association<reco::VertexCollection> > PVAsso_;
+    const edm::EDGetTokenT<edm::ValueMap<int> >            PVAssoQuality_;
+    const edm::EDGetTokenT<reco::VertexCollection>         PVOrigs_;
+    const edm::EDGetTokenT<reco::TrackCollection>          TKOrigs_;
+    const edm::EDGetTokenT< edm::ValueMap<float> >         PuppiWeight_;
+    const edm::EDGetTokenT<edm::ValueMap<reco::CandidatePtr> >    PuppiCandsMap_;
+    const edm::EDGetTokenT<std::vector< reco::PFCandidate >  >    PuppiCands_;
+    const edm::EDGetTokenT<edm::View<reco::CompositePtrCandidate> > SVWhiteList_;
+    
+    const double minPtForTrackProperties_;
+    // for debugging
+    float calcDxy(float dx, float dy, float phi) const {
+      return - dx * std::sin(phi) + dy * std::cos(phi);
+    }
+    float calcDz(reco::Candidate::Point p, reco::Candidate::Point v, const reco::Candidate &c) const {
+      return p.Z()-v.Z() - ((p.X()-v.X()) * c.px() + (p.Y()-v.Y())*c.py()) * c.pz()/(c.pt()*c.pt());
```

Comment:
Since we don't reconstruct any tracks going exactly down the beam pipe, we
never generate (good) tracks with pt == 0 so this cannot happen in this
module.

-L

(Sent from my Nexus 6)
On Aug 14, 2015 11:37 PM, "Carl Vuosalo" <notifications@github.com> wrote:

> In PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc
> <https://github.com/cms-sw/cmssw/pull/10620#discussion_r37121921>:
>
> > +    const edm::EDGetTokenT<edm::Association<reco::VertexCollection> > PVAsso_;
> > +    const edm::EDGetTokenT<edm::ValueMap<int> >            PVAssoQuality_;
> > +    const edm::EDGetTokenT<reco::VertexCollection>         PVOrigs_;
> > +    const edm::EDGetTokenT<reco::TrackCollection>          TKOrigs_;
> > +    const edm::EDGetTokenT< edm::ValueMap<float> >         PuppiWeight_;
> > +    const edm::EDGetTokenT<edm::ValueMap<reco::CandidatePtr> >    PuppiCandsMap_;
> > +    const edm::EDGetTokenT<std::vector< reco::PFCandidate >  >    PuppiCands_;
> > +    const edm::EDGetTokenT<edm::View<reco::CompositePtrCandidate> > SVWhiteList_;
> > +
> > +    const double minPtForTrackProperties_;
> > +    // for debugging
> > +    float calcDxy(float dx, float dy, float phi) const {
> > +      return - dx * std::sin(phi) + dy * std::cos(phi);
> > +    }
> > +    float calcDz(reco::Candidate::Point p, reco::Candidate::Point v, const reco::Candidate &c) const {
> > +      return p.Z()-v.Z() - ((p.X()-v.X()) * c.px() + (p.Y()-v.Y())*c.py()) * c.pz()/(c.pt()*c.pt());
>
> is c.pt() guaranteed to be greater than 0? If not, check for it to
> prevent division by zero.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10620/files#r37121921>.
>